i6_models/config.py

```diff
@@ -51,3 +54,28 @@ def _validate_types(self) -> None:
 
     def __post_init__(self) -> None:
         self._validate_types()
+
+
+ConfigType = TypeVar("ConfigType", bound=ModelConfiguration)
+ModuleType = TypeVar("ModuleType", bound=nn.Module)
+
+
+@dataclass
+class ModuleFactoryV1(Generic[ConfigType, ModuleType]):
+    """
+    Dataclass for a combination of a Subassembly/Part and the corresponding configuration.
+    Also provides a function to construct the corresponding object through this dataclass
+    """
+
+    module_class: Type[ModuleType]
+    cfg: ConfigType
+
+    def __call__(self) -> ModuleType:
+        """Constructs an instance of the given module class"""
+        return self.module_class(self.cfg)
+
+    def __post_init__(self) -> None:
+        parameters = signature(self.module_class).parameters.values()
+        assert len(parameters) == 1
+        cfg_parameter = next(iter(parameters))
+        typeguard.check_type(self.cfg, cfg_parameter.annotation)
```

Comment:
Result of offline discussion: Solving the issue via type-hinting seems to be either not possible or overly complicated. Instead have a more sophisticated type-check here that should cover all cases of how `module_class` can be defined.
```suggestion
        # Check typing of module_class and cfg, i.e. make sure that "self.module_class(self.cfg)" is a valid call.
        parameters = inspect.signature(self.module_class).parameters.values()
        assert len(parameters) >= 1
        parameter_iter = iter(parameters)

        # 1. Check that the first parameter is either not annotated or the annotation matches the type of self.cfg
        cfg_parameter = next(parameter_iter)
        if cfg_parameter.annotation is not inspect.Parameter.empty:
            typeguard.check_type(self.cfg, cfg_parameter.annotation)
            
        # 2. Check that all other parameters have default values
        for parameter in parameter_iter:
            assert parameter.default is not inspect.Parameter.empty
```